### PR TITLE
icmp6-util: several cleanups and fixlets

### LINKED ISSUE #29050: systemd-networkd does not reply to RS from unspecified address
### systemd version the issue has been seen with

253.6

### Used distribution

Gentoo

### Linux kernel version used

6.1.51-longterm

### CPU architectures issue was seen on

x86_64

### Component

systemd-networkd

### Expected behaviour you didn't see

QEMU VM attempting HTTP boot over IPv6 from EDK2 (edk2-stable202302-for-qemu) BIOS are emitting IPv6 router solicitation from `::` source address (unspecified address) as per following pcap extract (full pcap available on https://mega.nz/file/11w1GBJB#M_GVOIoJLMJch404wpXY5VwkvxAOELzyVJpLr0RfW0c):

```
Frame 5: 62 bytes on wire (496 bits), 62 bytes captured (496 bits)
    Encapsulation type: Ethernet (1)
    Arrival Time: Sep  3, 2023 22:23:44.296580000 IST
    [Time shift for this packet: 0.000000000 seconds]
    Epoch Time: 1693776224.296580000 seconds
    [Time delta from previous captured frame: 0.956635000 seconds]
    [Time delta from previous displayed frame: 0.956635000 seconds]
    [Time since reference or first frame: 0.961068000 seconds]
    Frame Number: 5
    Frame Length: 62 bytes (496 bits)
    Capture Length: 62 bytes (496 bits)
    [Frame is marked: False]
    [Frame is ignored: False]
    [Protocols in frame: eth:ethertype:ipv6:icmpv6]
Ethernet II, Src: 52:54:00:1e:33:35, Dst: 33:33:00:00:00:02
    Destination: 33:33:00:00:00:02
        Address: 33:33:00:00:00:02
        .... ..1. .... .... .... .... = LG bit: Locally administered address (this is NOT the factory default)
        .... ...1 .... .... .... .... = IG bit: Group address (multicast/broadcast)
    Source: 52:54:00:1e:33:35
        Address: 52:54:00:1e:33:35
        .... ..1. .... .... .... .... = LG bit: Locally administered address (this is NOT the factory default)
        .... ...0 .... .... .... .... = IG bit: Individual address (unicast)
    Type: IPv6 (0x86dd)
Internet Protocol Version 6, Src: ::, Dst: ff02::2
    0110 .... = Version: 6
    .... 0000 0000 .... .... .... .... .... = Traffic Class: 0x00 (DSCP: CS0, ECN: Not-ECT)
        .... 0000 00.. .... .... .... .... .... = Differentiated Services Codepoint: Default (0)
        .... .... ..00 .... .... .... .... .... = Explicit Congestion Notification: Not ECN-Capable Transport (0)
    .... 0000 0000 0000 0000 0000 = Flow Label: 0x00000
    Payload Length: 8
    Next Header: ICMPv6 (58)
    Hop Limit: 255
    Source Address: ::
    Destination Address: ff02::2
Internet Control Message Protocol v6
    Type: Router Solicitation (133)
    Code: 0
    Checksum: 0x7bb8 [correct]
    [Checksum Status: Good]
    Reserved: 00000000
```

According to https://datatracker.ietf.org/doc/html/rfc4861#section-6.1.1 and https://datatracker.ietf.org/doc/html/rfc4861#section-6.2.6, such solicitation should be honoured, however systemd-networkd ignore them with the following error:

```
Sep 03 21:23:44 cook.local systemd-networkd[38040]: br0.128: RADV: Received RS from non-link-local address ::5d6d:bf15:327f:0. Ignoring
```

I've applied to following patch on my local router to allow systemd to reply to IPv6 router solicitation when source IP is part of link-local address space and also unspecified address:

```
diff --git a/src/libsystemd-network/icmp6-util.c b/src/libsystemd-network/icmp6-util.c
index 00a59291a1c4..2048532d18a2 100644
--- a/src/libsystemd-network/icmp6-util.c
+++ b/src/libsystemd-network/icmp6-util.c
@@ -178,7 +178,7 @@ int icmp6_receive(int fd, void *buffer, size_t size, struct in6_addr *ret_dst,
             sa.in6.sin6_family == AF_INET6)  {

                 addr = sa.in6.sin6_addr;
-                if (!in6_addr_is_link_local(&addr))
+                if (!in6_addr_is_link_local(&addr) && !in6_addr_is_null(&addr))
                         return -EADDRNOTAVAIL;

         } else if (msg.msg_namelen > 0)
```

This patch has the expected result and systemd-networkd now reply to such IPv6 router solicitation:

```
Sep 04 01:30:42 cook.local systemd-networkd[139302]: br0.128: RADV: Sent solicited Router Advertisement to ::
```

However I'm quite unsure the impact this could lead to the rest of code and/or this should be applied in other decision path made by systemd-networkd

### Unexpected behaviour you saw

systemd-networkd should reply to IPv6 router solicitation from `::` source address (unspecified address)

### Steps to reproduce the problem

1. Enable IPv6 routing with systemd
2. Launch VM with QEmu booting on HTTP IPv6 from BIOS directly

### Additional program output to the terminal or log subsystem illustrating the issue

_No response_

### PR DESCRIPTION
**Pull request by yuwata**
Fixes #29050.

(Please backport the first three commits into stable.)

### CODE REVIEW

File: src/libsystemd-network/fuzz-ndisc-rs.c [hunk 1]
```diff
@@ -24,12 +24,17 @@ int icmp6_bind_router_advertisement(int index) {
         return -ENOSYS;
 }
 
-int icmp6_receive(int fd, void *iov_base, size_t iov_len,
-                  struct in6_addr *dst, triple_timestamp *timestamp) {
+int icmp6_receive(
+                int fd,
+                void *iov_base,
+                size_t iov_len,
+                struct in6_addr *ret_sendor,
```
Comment (bluca) [inline src/libsystemd-network/fuzz-ndisc-rs.c]: "sender"?

**Review by bluca — APPROVED** [review]

File: src/libsystemd-network/fuzz-ndisc-rs.c (hunk 1 above)
Comment (yuwata) [reply src/libsystemd-network/fuzz-ndisc-rs.c]: Ouch.

**Comment by yuwata** [comment]
@bluca @ssahani Thank you for the review. Typo is fixed. Upgrading the green label.

File: src/libsystemd-network/icmp6-util-unix.c [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+                return 0;
+
+        return send_ra_function(0);
+}
+
+int icmp6_receive(
+                int fd,
+                void *iov_base,
+                size_t iov_len,
+                struct in6_addr *ret_sendor,
```
Comment (mrc0mmand) [inline src/libsystemd-network/icmp6-util-unix.c]:
```suggestion
                struct in6_addr *ret_sender,
```

File: src/libsystemd-network/icmp6-util-unix.c [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+                triple_timestamp *ret_timestamp) {
+
+        assert_se(read (fd, iov_base, iov_len) == (ssize_t) iov_len);
+
+        if (ret_timestamp)
+                triple_timestamp_get(ret_timestamp);
+
+        if (ret_sendor)
+                *ret_sendor = dummy_link_local;
```
Comment (mrc0mmand) [inline src/libsystemd-network/icmp6-util-unix.c]:
```suggestion
        if (ret_sender)
                *ret_sender = dummy_link_local;
```

File: src/libsystemd-network/test-ndisc-rs.c [hunk 4]
```diff
@@ -41,7 +37,8 @@ static void router_dump(sd_ndisc_router *rt) {
         assert_se(rt);
 
         log_info("--");
-        assert_se(sd_ndisc_router_get_address(rt, &addr) == -ENODATA);
+        assert_se(sd_ndisc_router_get_address(rt, &addr) >= 0);
+        log_info("Sendor: %s", IN6_ADDR_TO_STRING(&addr));
```
Comment (mrc0mmand) [inline src/libsystemd-network/test-ndisc-rs.c]:
```suggestion
        log_info("Sender: %s", IN6_ADDR_TO_STRING(&addr));
```

**Review by mrc0mmand — COMMENTED** [review]
Spotted the typo in a couple more places.

File: src/libsystemd-network/test-ndisc-rs.c (hunk 4 above)
Comment (yuwata) [reply src/libsystemd-network/test-ndisc-rs.c]: Ouch ouch ouch. My fingers learned the wrong spell...

**Pull request merged**